pkg/operator/webhooksupportabilitycontroller/degraded_webhook.go

```diff
@@ -0,0 +1,120 @@
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/apimachinery/pkg/util/runtime"
+)
+
+// webhookInfo generically represents a webhook
+type webhookInfo struct {
+	Name     string
+	Service  *serviceReference
+	CABundle []byte
+}
+
+// serviceReference generically represents a service reference
+type serviceReference struct {
+	Namespace string
+	Name      string
+	Port      *int32
+}
+
+// updateWebhookConfigurationDegraded updates the condition specified after
+// checking that the services associated with the specified webhooks exist
+// and have at least one ready endpoint.
+func (c *webhookSupportabilityController) updateWebhookConfigurationDegraded(ctx context.Context, condition operatorv1.OperatorCondition, webhookInfos []webhookInfo) v1helpers.UpdateStatusFunc {
+	var serviceMsgs []string
+	var tlsMsgs []string
+	for _, webhook := range webhookInfos {
+		if webhook.Service != nil {
+			err := c.assertService(webhook.Service)
+			if err != nil {
+				serviceMsgs = append(serviceMsgs, fmt.Sprintf("%s: %s", webhook.Name, err))
+				continue
+			}
+			err = c.assertConnect(ctx, webhook.Service, webhook.CABundle)
+			if err != nil {
+				tlsMsgs = append(tlsMsgs, fmt.Sprintf("%s: %s", webhook.Name, err))
+				continue
+			}
+		}
+	}
+
+	svc, tls := len(serviceMsgs) > 0, len(tlsMsgs) > 0
+	switch {
+	case svc && tls:
+		condition.Reason = WebhookServiceNotReadyReason
+		condition.Status = operatorv1.ConditionTrue
+	case svc:
+		condition.Reason = WebhookServiceNotFoundReason
+		condition.Status = operatorv1.ConditionTrue
+	case tls:
+		condition.Reason = WebhookServiceConnectionErrorReason
+		condition.Status = operatorv1.ConditionTrue
+	default:
+		condition.Reason = ""
+		condition.Status = operatorv1.ConditionFalse
+	}
+	msgs := append(serviceMsgs, tlsMsgs...)
+	sort.Strings(msgs)
+	condition.Message = strings.Join(msgs, "\n")
+
+	return v1helpers.UpdateConditionFn(condition)
+}
+
+// assertService checks that the referenced service resource exists.
+func (c *webhookSupportabilityController) assertService(reference *serviceReference) error {
+	_, err := c.serviceLister.Services(reference.Namespace).Get(reference.Name)
+	if err != nil {
+		return fmt.Errorf("unable to find find service %s.%s: %v", reference.Name, reference.Namespace, err)
+	}
+	return nil
+}
+
+// assertConnect performs a dns lookup of service, opens a tcp connection, and performs a tls handshake.
+func (c *webhookSupportabilityController) assertConnect(ctx context.Context, reference *serviceReference, caBundle []byte) error {
+	host := reference.Name + "." + reference.Namespace + ".svc"
+	port := "443"
+	if reference.Port != nil {
+		port = fmt.Sprintf("%d", *reference.Port)
+	}
+	rootCAs := x509.NewCertPool()
+	if len(caBundle) > 0 {
+		rootCAs.AppendCertsFromPEM(caBundle)
+	}
+	var err error
+	// retry up to 3 times on error
+	for i := 0; i < 3; i++ {
+		select {
+		case <-ctx.Done():
+			return nil
```

Comment:
should we return `ctx.Err()` ?

Reply:
No. Whatever caused the context to be cancelled should not be affecting the degraded condition.

Reply:
in that case, I think we could update the `library-go` and pass a context to `v1helpers.UpdateStatus` function - it won't update the condition if the `ctx` expired

Reply:
I check for a cancelled context before the call to `UpdateStatus`. I'll bump `library-go` in a subsequent PR.